library/src/main/java/com/datatorrent/lib/state/BucketedState.java

```diff
@@ -0,0 +1,101 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.datatorrent.lib.state;
+
+import java.util.concurrent.Future;
+
+import com.datatorrent.netlet.util.Slice;
+
+/**
+ * A state where keys are grouped in buckets.
+ */
+public interface BucketedState
+{
+  /**
+   * Sets the value of the key in bucket identified by bucketId.
+   *
+   * @param bucketId identifier of the bucket.
+   * @param key     key
+   * @param value   value
+   */
+  void put(long bucketId, Slice key, Slice value);
```

Comment:
Is bucketId an artifact of the implementation? The key is fine as it has a manifestation for the user within the application. This way of specification puts an implicit 2 level hierarchy. Questions arise why not 3 levels?